src/world/map.c

```diff
@@ -4503,77 +4503,91 @@ void map_extend_boundary_surface()
-		case MAP_ELEMENT_TYPE_SURFACE:
-			mapElement->base_height = 2;
-			mapElement->clearance_height = 2;
-			mapElement->properties.surface.slope = 0;
-			mapElement->properties.surface.terrain = 0;
-			mapElement->properties.surface.grass_length = 1;
-			mapElement->properties.surface.ownership = 0;
-			if (!map_element_is_last_for_tile(mapElement++))
-				goto next_element;
-
-			return;
-		case MAP_ELEMENT_TYPE_ENTRANCE:
-			viewport_interaction_remove_park_entrance(mapElement, x, y);
-			break;
-		case MAP_ELEMENT_TYPE_FENCE:
-			gGameCommandErrorTitle = STR_CANT_REMOVE_THIS;
-			game_do_command(
+	rct_map_element *element = *elementPtr;
+	switch (map_element_get_type(element)) {
+	case MAP_ELEMENT_TYPE_SURFACE:
+		element->base_height = 2;
+		element->clearance_height = 2;
+		element->properties.surface.slope = 0;
+		element->properties.surface.terrain = 0;
+		element->properties.surface.grass_length = 1;
+		element->properties.surface.ownership = 0;
+		// Because this element is not completely removed, the pointer must be updated muanually
+		// The rest of the elements are removed from the array, so the pointer doesn't need to be updated.
+		(*elementPtr)++;
+		break;
+	case MAP_ELEMENT_TYPE_ENTRANCE:
+		viewport_interaction_remove_park_entrance(element, x, y);
+		break;
+	case MAP_ELEMENT_TYPE_FENCE:
+		gGameCommandErrorTitle = STR_CANT_REMOVE_THIS;
+		game_do_command(
 				x,
 				GAME_COMMAND_FLAG_APPLY,
 				y,
-				(mapElement->type & MAP_ELEMENT_DIRECTION_MASK) | (mapElement->base_height << 8),
+				(element->type & MAP_ELEMENT_DIRECTION_MASK) | (element->base_height << 8),
 				GAME_COMMAND_REMOVE_FENCE,
 				0,
 				0
-			);
-			break;
-		case MAP_ELEMENT_TYPE_SCENERY_MULTIPLE:
-			gGameCommandErrorTitle = STR_CANT_REMOVE_THIS;
-			game_do_command(
+		);
+		break;
+	case MAP_ELEMENT_TYPE_SCENERY_MULTIPLE:
+		gGameCommandErrorTitle = STR_CANT_REMOVE_THIS;
+		game_do_command(
 				x,
-				(GAME_COMMAND_FLAG_APPLY) | ((mapElement->type & MAP_ELEMENT_DIRECTION_MASK) << 8),
+				(GAME_COMMAND_FLAG_APPLY) | ((element->type & MAP_ELEMENT_DIRECTION_MASK) << 8),
 				y,
-				(mapElement->base_height) | (((mapElement->properties.scenerymultiple.type >> 8) >> 2) << 8),
+				(element->base_height) | (((element->properties.scenerymultiple.type >> 8) >> 2) << 8),
 				GAME_COMMAND_REMOVE_LARGE_SCENERY,
 				0,
 				0
-			);
-			break;
-		case MAP_ELEMENT_TYPE_BANNER:
-			gGameCommandErrorTitle = STR_CANT_REMOVE_THIS;
-			game_do_command(
+		);
+		break;
+	case MAP_ELEMENT_TYPE_BANNER:
+		gGameCommandErrorTitle = STR_CANT_REMOVE_THIS;
+		game_do_command(
 				x,
 				GAME_COMMAND_FLAG_APPLY,
 				y,
-				(mapElement->base_height) | ((mapElement->properties.banner.position & 3) << 8),
+				(element->base_height) | ((element->properties.banner.position & 3) << 8),
 				GAME_COMMAND_REMOVE_BANNER,
 				0,
 				0
-			);
-			break;
-		default:
-			map_element_remove(mapElement);
-			break;
+		);
+		break;
+	default:
```

Comment:
This will cause rides to break if it's ever called on a map with rides. It should really delete the whole ride if any bit is being removed.